File: code/python/1482.py
#https://leetcode.com/problems/minimum-number-of-days-to-make-m-bouquets/
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        def countBlooms(day):
            i = count = 0
            for e in bloomDay:
                if e<= day:
                    i+=1
                    if i == k:
                        count+=1
                        if count == m: return True
                        i = 0
                else:
                    i = 0
            return False
        
        target = m * k
        if target > len(bloomDay): return -1
        l = min(bloomDay)
        r = max(bloomDay)
        while l < r:
            mid = (l + r) // 2
            if countBlooms(mid):
                r = mid
            else:
                l = mid + 1
        return int(l)

File: code/cpp/1482.cpp
//https://leetcode.com/problems/minimum-number-of-days-to-make-m-bouquets/
class Solution {
public:
    bool countBlooms(vector<int>& bloomDay, int m, int k, int day) {
        size_t i = 0, count = 0;
        for (int e : bloomDay) {
            if (e<= day) {
                i++;
                if (i == k) {
                    count++;
                    if (count == m) return true;
                    i = 0;
                }
            }
            else
                i = 0;
        }
        return false;
    }

    int minDays(vector<int>& bloomDay, int m, int k) {
        size_t target = (size_t)m * k;
        if (target > bloomDay.size()) return -1;
        size_t l = 1, mid, r = 1e9;
        while (l < r) {
            mid = (l + r) / 2;
            if (countBlooms(bloomDay, m, k, mid))
                r = mid;
            else
                l = mid + 1;
        }
        return l;
    }
};

static int speedup=[](){
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	return 0;
}();
